src/intellisense/langCompletionProvider.ts

```diff
@@ -0,0 +1,457 @@
+        const wordRange = document.getWordRangeAtPosition(position, this.contextRegex);
+        if (wordRange) {
+            const match = document.getText(wordRange).match(this.contextRegex);
+            const {path, type} = this.parseMatch(match as RegExpMatchArray);
+            return this.getCompletionItems(document.uri, path, type);
+        }
+        return Promise.resolve([]);
+    }
+
+    provideDocumentLinks(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.ProviderResult<vscode.DocumentLink[]> {
+        return this.getDocumentLinks(document.uri, document);
+    }
+
+    get triggers(): vscode.Disposable[] {
+        const selector = {...this.selector, pattern: '**/*.{tex,txt}'};
+        return [
+            // register completion provider
+            vscode.languages.registerCompletionItemProvider(selector, this, '{', '/'),
+            // register document link provider
+            vscode.languages.registerDocumentLinkProvider(selector, this),
+        ];
+    }
+}
+
+export class ReferenceCompletionProvider extends IntellisenseProvider implements vscode.CompletionItemProvider {
+    protected readonly contextPrefix = [
+        // group 0: reference
+        ['\\w*ref'],
+        // group 1: citation
+        ['cite'],
+    ];
+
+    constructor(vfsm:RemoteFileSystemProvider, private readonly symbolProvider:DocSymbolProvider){
+        super(vfsm);
+    }
+
+    private parseMatch(match: RegExpMatchArray) {
+        const keywords = match.slice(1, -1);
+        const index = keywords.findIndex(x => x!==undefined);
+        const _match = (match.at(-1) as string).split(/(.*),([^,]*)/);
+        const partial = _match.length===1 ? _match[0] : _match[2];
+        return {index, partial};
+    }
+
+    private async getCompletionItems(uri:vscode.Uri, idx: number, partial:string): Promise<vscode.CompletionItem[]> {
+        const vfs = await this.vfsm.prefetch(uri);
+        switch (idx) {
+            case 0: // group 0: reference
+                const res = await vfs.metadata();
+                if (res===undefined) { return []; };
+                return Object.values(res).map(({labels}) => {
+                    return labels.map(label => new vscode.CompletionItem(label, vscode.CompletionItemKind.Reference));
+                }).flat();
+            case 1: // group 1: citation
+                let items = await this.getReferenceCompletionItemsFromBib(vfs);
+                items = items.length!==0 ? items : await this.getReferenceCompletionItemsFromBbl(vfs); //fallback option
+                return items;
+            default:
+                return [];
+        }
+    }
+
+    private async getReferenceCompletionItemsFromBib(vfs: VirtualFileSystem): Promise<vscode.CompletionItem[]> {
+        const bibList:string[] = this.symbolProvider.getBibList() ?? [];
+        const bibPaths = bibList.flatMap(
+            path => (path.split(',') ?? [])
```

Comment:
@QianrenLi why the path is joined with `,`. I cannot find the hint anywhere.

Reply:
Sometimes, the path included in the macro `\bibgraphy` can be multiple file like `{ieee,reference}'.  The split is used to sperate them up.